paderwasn/rir_simulation/rir_simulation.py

```diff
@@ -0,0 +1,525 @@
+    msg = "Invalid source directivity specified"
+    assert src_directivity in pattern_dispatcher.keys(), msg
+
+    msg = "Only simulation of microphone pairs is supported"
+    assert len(np.array(sensor_pos).T) == 2, msg
+
+    if np.linalg.norm(np.diff(sensor_pos, axis=1)) > 0.1:
+        warnings.warn("The first microphone determines the directivity factor \
+            for the other. This assumption may not hold for larger microphone \
+            spacings where larger differences for alpha may be possible.")
+
+    # simulate first "max_order_ism" reflections with pyroomacoustics
+    e_absorption, _ = pra.inverse_sabine(t60, room_dimensions)
+    room = pra.ShoeBox(
+        room_dimensions, fs=sample_rate, materials=pra.Material(e_absorption),
+        max_order=max_order_ism, use_rand_ism=True, max_rand_disp=0.001)
+
+    directivity_src = CardioidFamily(
+        orientation=DirectionVector(azimuth=azimuth_directivity,
+                                    colatitude=colatitude_directivity,
+                                    degrees=True),
+        pattern_enum=pattern_dispatcher[src_directivity]
+    )
+    room.add_source(src_pos, directivity=directivity_src)
+    room.add_microphone(sensor_pos)
+    room.compute_rir()
+
+    # Apply high-pass filter
+    rirs_ism = np.zeros((2, rir_len))
+    for ch_id, rir_ism in enumerate(room.rir):
+        assert np.any(rir_ism != 0)
+        rir_ism = high_pass(rir_ism[0])
+        rirs_ism[ch_id, :len(rir_ism)] = rir_ism
+
+    it_cnt = 0
+    any_result_none = False
+    change_alpha = False
+    rirs_stochastic = np.zeros((2, rir_len))
+    alpha_values = np.zeros(2)
+
+    # calculate stochastic part until possible solution found
+    while (np.abs(alpha_values[0] - alpha_values[1]) > alpha_diff
+           or not all(np.any(rirs_stochastic, axis=-1))):
+        if it_cnt == 0:
+            alpha = np.random.uniform(*directivity_factor_range)
+        elif any_result_none and it_cnt < max_retries:
+            # rejection sampling for 'max_retries' iterations
+            alpha = np.random.uniform(*directivity_factor_range)
+        else:
+            # if maximum possible factor calculated, choose minimum
+            # to have same factor for the different microphones
+            alpha = min(alpha_values)
+
+        it_cnt += 1
+        if it_cnt > max_retries:
+            change_alpha = True
+
+        for ch, rir_ism in enumerate(rirs_ism):
+            # calculate directivity attenuation
+            delta_pos = sensor_pos[:, ch] - np.asarray(src_pos)
+            directivity_attenuation =\
+                calc_directivity_attenuation(delta_pos, azimuth_directivity,
+                                             colatitude_directivity,
+                                             src_directivity)
+
+            mic_pos = sensor_pos.T[ch]
+            distance = np.linalg.norm(mic_pos - src_pos)
+
+            # Note: the first microphone determines the directivity factor
```

Comment:
Note: the first microphone determines the directivity factor since the directivity factor should be approximately the same for both microphones if the inter-microphone distance is assumes to be small. For larger  inter-microphone distance, the directivity factor might be different for both microphones.